tlatools/org.lamport.tlatools/src/tlc2/REPL.java

```diff
@@ -0,0 +1,166 @@
+    Path replTempDir;
+
+    public REPL(Path tempDir) {
+        replTempDir = tempDir;
+    }
+
+    public void setSpecFile(final File pSpecFile) {
+        specFile = pSpecFile;
+    }
+
+    /**
+     * Evaluate the given string input as a TLA+ expression.
+     *
+     * @return the pretty printed result of the evaluation or an empty string if there was an error.
+     */
+    public String processInput(String evalExpr) {
+
+        // The modules we will extend in the REPL environment.
+        String moduleExtends = "Naturals,Reals,Sequences,Bags,FiniteSets,TLC";
+        if (specFile != null) {
+            String mainModuleName = specFile.getName().replaceFirst(TLAConstants.Files.TLA_EXTENSION + "$", "");
+            moduleExtends += ("," + mainModuleName);
+        }
+
+        File tempFile, configFile;
+        try {
+
+            // We want to place the spec files used by REPL evaluation into the temporary directory.
+            tempFile = new File(replTempDir.toString(), REPL_SPEC_NAME + TLAConstants.Files.TLA_EXTENSION);
+            configFile = new File(replTempDir.toString(), REPL_SPEC_NAME + TLAConstants.Files.CONFIG_EXTENSION);
+
+            // Create the config file.
+            BufferedWriter cfgWriter = new BufferedWriter(new FileWriter(configFile.getAbsolutePath(), false));
+            cfgWriter.append("INIT replinit");
+            cfgWriter.newLine();
+            cfgWriter.append("NEXT replnext");
+            cfgWriter.newLine();
+            cfgWriter.close();
+
+            // Create the spec file lines.
+            ArrayList<String> lines = new ArrayList<String>();
+            String replValueVarName = "replvalue";
+            lines.add("---- MODULE tlarepl ----");
+            lines.add("EXTENDS " + moduleExtends);
+            lines.add("VARIABLE replvar");
+            // Dummy Init and Next predicates.
+            lines.add("replinit == replvar = 0");
+            lines.add("replnext == replvar' = 0");
+            // The expression to evaluate.
+            lines.add(replValueVarName + " == " + evalExpr);
+            lines.add("====");
+
+            // Write out the spec file.
+            BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile.getAbsolutePath(), false));
+            for (String line : lines) {
+                writer.append(line);
+                writer.newLine();
+            }
+            writer.close();
+
+            // Avoid sending log messages to stdout and reset the messages recording.
+            ToolIO.setMode(ToolIO.TOOL);
+            ToolIO.reset();
+
+            try {
+                // We placed the REPL spec files into a temporary directory, so, we add this temp directory
+                // path to the filename resolver used by the Tool.
+                SimpleFilenameToStream resolver = new SimpleFilenameToStream(replTempDir.toAbsolutePath().toString());
+                Tool tool = new FastTool(REPL_SPEC_NAME, REPL_SPEC_NAME, resolver);
```

Comment:
What makes `FastTool` faster is that it doesn't collect information for an error stack if an expression fails to evaluate.  I assume a REPL benefits from tlc2.tool.impl.CallStackTool that collects an error stack.

Reply:
Ok, good to know. I briefly tried out `CallStackTool` in place of `FastTool` here and it didn't seem noticeably slower in terms of REPL latency. Using `CallStackTool` might be necessary to improve error reporting, but I was hoping to defer that work to follow up patches. Is that acceptable to you?